upb/msg_test.cc

```diff
@@ -0,0 +1,101 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Google LLC nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL Google LLC BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "gtest/gtest.h"
+#include "src/google/protobuf/test_messages_proto3.upb.h"
+#include "upb/def.hpp"
+#include "upb/json_decode.h"
+#include "upb/json_encode.h"
+#include "upb/msg_test.upb.h"
+#include "upb/msg_test.upbdefs.h"
+#include "upb/upb.hpp"
+
+void VerifyMessage(const upb_test_TestExtensions *ext_msg) {
+  EXPECT_TRUE(upb_test_TestExtensions_has_optional_int32_ext(ext_msg));
+  // EXPECT_FALSE(upb_test_TestExtensions_Nested_has_optional_int32_ext(ext_msg));
+  EXPECT_TRUE(upb_test_has_optional_msg_ext(ext_msg));
+
+  EXPECT_EQ(123, upb_test_TestExtensions_optional_int32_ext(ext_msg));
+  const protobuf_test_messages_proto3_TestAllTypesProto3 *ext_submsg =
+      upb_test_optional_msg_ext(ext_msg);
+  EXPECT_TRUE(ext_submsg != nullptr);
+  EXPECT_EQ(456,
+            protobuf_test_messages_proto3_TestAllTypesProto3_optional_int32(
+                ext_submsg));
+}
+
+TEST(MessageTest, Extensions) {
+  upb::Arena arena;
+  upb_test_TestExtensions *ext_msg = upb_test_TestExtensions_new(arena.ptr());
+
+  EXPECT_FALSE(upb_test_TestExtensions_has_optional_int32_ext(ext_msg));
+  // EXPECT_FALSE(upb_test_TestExtensions_Nested_has_optional_int32_ext(ext_msg));
+  EXPECT_FALSE(upb_test_has_optional_msg_ext(ext_msg));
+
+  upb::SymbolTable symtab;
+  upb::MessageDefPtr m(upb_test_TestExtensions_getmsgdef(symtab.ptr()));
+  EXPECT_TRUE(m.ptr() != nullptr);
+
+  std::string json = R"(
```

Comment:
I think if you write this as

R"json(

)json"

then clang-format will format the internal stuff too

Reply:
Done.  (it didn't make a difference for me, but I might not have a new enough version of clang-format.